statistics/row_sampler.go

```diff
@@ -0,0 +1,306 @@
+	}
+	ctx := context.TODO()
+	chk := s.RecordSet.NewChunk()
+	it := chunk.NewIterator4Chunk(chk)
+	for {
+		err := s.RecordSet.Next(ctx, chk)
+		if err != nil {
+			return nil, err
+		}
+		if chk.NumRows() == 0 {
+			return collector, nil
+		}
+		collector.Count += int64(chk.NumRows())
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			datums := RowToDatums(row, s.RecordSet.Fields())
+			for i, val := range datums {
+				// For string values, we use the collation key instead of the original value.
+				if s.Collators[i] != nil && !val.IsNull() {
+					decodedVal, err := tablecodec.DecodeColumnValue(val.GetBytes(), s.ColsFieldType[i], s.Sc.TimeZone)
+					if err != nil {
+						return nil, err
+					}
+					decodedVal.SetBytesAsString(s.Collators[i].Key(decodedVal.GetString()), decodedVal.Collation(), uint32(decodedVal.Length()))
+					encodedKey, err := tablecodec.EncodeValue(s.Sc, nil, decodedVal)
+					if err != nil {
+						return nil, err
+					}
+					val.SetBytes(encodedKey)
+				}
+			}
+			err := collector.collectColumns(s.Sc, datums)
+			if err != nil {
+				return nil, err
+			}
+			err = collector.collectColumnGroups(s.Sc, datums, s.ColGroups)
+			if err != nil {
+				return nil, err
+			}
+			weight := s.Rng.Int63()
+			collector.sampleRow(datums, weight)
+		}
+	}
+}
+
+func (s *RowSampleCollector) collectColumns(sc *stmtctx.StatementContext, cols []types.Datum) error {
+	for i, col := range cols {
+		if col.IsNull() {
+			s.NullCount[i]++
+		}
+		s.TotalSizes[i] += int64(len(col.GetBytes())) - 1
+		// Minus one is to remove the flag byte.
+		err := s.FMSketches[i].InsertValue(sc, col)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *RowSampleCollector) collectColumnGroups(sc *stmtctx.StatementContext, cols []types.Datum, colGroups [][]int64) error {
+	colLen := len(cols)
+	datumBuffer := make([]types.Datum, 0, len(cols))
+	for i, group := range colGroups {
+		datumBuffer = datumBuffer[:0]
+		allNull := true
+		for _, c := range group {
+			datumBuffer = append(datumBuffer, cols[c])
+			allNull = allNull && cols[c].IsNull()
+			s.TotalSizes[colLen+i] += int64(len(cols[c].GetBytes())) - 1
+		}
+		err := s.FMSketches[colLen+i].InsertRowValue(sc, datumBuffer)
+		if err != nil {
+			return err
+		}
+		if allNull {
+			s.NullCount[colLen+i]++
+		}
+	}
+	return nil
+}
+
+func (s *RowSampleCollector) sampleRow(cols []types.Datum, weight int64) {
+	if len(s.Samples) < s.MaxSampleSize {
+		newCols := make([]types.Datum, len(cols))
+		for i := range cols {
+			cols[i].Copy(&newCols[i])
+		}
+		s.Samples = append(s.Samples, &RowSampleItem{
+			Columns: newCols,
+			Weight:  weight,
+		})
+		if len(s.Samples) == s.MaxSampleSize {
+			heap.Init(&s.Samples)
+		}
+		return
+	}
+	if s.Samples[0].Weight < weight {
+		for i := range cols {
+			cols[i].Copy(&s.Samples[0].Columns[i])
+		}
+		s.Samples[0].Weight = weight
+		heap.Fix(&s.Samples, 0)
+	}
+}
```

Comment:
The samples collected does not maintain their original orders, that would lead to wrong correlation computed in analyze.

Reply:
We can sort them later by the given orders. When collecting the primary key's statistics.